File: Android/APIExample/app/src/main/java/io/agora/beautyapi/faceunity/utils/APIReporter.kt
package io.agora.beautyapi.faceunity.utils

import android.util.Log
import io.agora.rtc2.Constants
import io.agora.rtc2.RtcEngine
import org.json.JSONObject

enum class APIType(val value: Int) {
    KTV(1),             // K歌
    CALL(2),            // 呼叫连麦
    BEAUTY(3),          // 美颜
    VIDEO_LOADER(4),    // 秒开秒切
    PK(5),              // 团战
    VIRTUAL_SPACE(6),   //
    SCREEN_SPACE(7),    // 屏幕共享
    AUDIO_SCENARIO(8)   // 音频
}

enum class ApiEventType(val value: Int) {
    API(0),
    COST(1),
    CUSTOM(2)
}

object ApiEventKey {
    const val TYPE = "type"
    const val DESC = "desc"
    const val API_VALUE = "apiValue"
    const val TIMESTAMP = "ts"
    const val EXT = "ext"
}

object ApiCostEvent {
    const val CHANNEL_USAGE = "channelUsage"                 //频道使用耗时
    const val FIRST_FRAME_ACTUAL = "firstFrameActual"        //首帧实际耗时
    const val FIRST_FRAME_PERCEIVED = "firstFramePerceived"  //首帧感官耗时
}

class APIReporter(
    private val type: APIType,
    private val version: String,
    private val rtcEngine: RtcEngine
) {
    private val tag = "APIReporter"
    private val messageId = "agora:scenarioAPI"
    private val durationEventStartMap = HashMap<String, Long>()
    private val category = "${type.value}_Android_$version"

    init {
        configParameters()
    }

    // 上报普通场景化API
    fun reportFuncEvent(name: String, value: Map<String, Any>, ext: Map<String, Any>) {
        Log.d(tag, "reportFuncEvent: $name value: $value ext: $ext")
        val eventMap = mapOf(ApiEventKey.TYPE to ApiEventType.API.value, ApiEventKey.DESC to name)
        val labelMap = mapOf(ApiEventKey.API_VALUE to value, ApiEventKey.TIMESTAMP to getCurrentTs(), ApiEventKey.EXT to ext)
        val event = convertToJSONString(eventMap) ?: ""
        val label = convertToJSONString(labelMap) ?: ""
        rtcEngine.sendCustomReportMessage(messageId, category, event, label, 0)
    }

    fun startDurationEvent(name: String) {
        Log.d(tag, "startDurationEvent: $name")
        durationEventStartMap[name] = getCurrentTs()
    }

    fun endDurationEvent(name: String, ext: Map<String, Any>) {
        Log.d(tag, "endDurationEvent: $name")
        val beginTs = durationEventStartMap[name] ?: return
        durationEventStartMap.remove(name)
        val ts = getCurrentTs()
        val cost = (ts - beginTs).toInt()

        innerReportCostEvent(ts, name, cost, ext)
    }

    // 上报耗时打点信息
    fun reportCostEvent(name: String, cost: Int, ext: Map<String, Any>) {
        durationEventStartMap.remove(name)
        innerReportCostEvent(
            ts = getCurrentTs(),
            name = name,
            cost = cost,
            ext = ext
        )
    }

    // 上报自定义信息
    fun reportCustomEvent(name: String, ext: Map<String, Any>) {
        Log.d(tag, "reportCustomEvent: $name ext: $ext")
        val eventMap = mapOf(ApiEventKey.TYPE to ApiEventType.CUSTOM.value, ApiEventKey.DESC to name)
        val labelMap = mapOf(ApiEventKey.TIMESTAMP to getCurrentTs(), ApiEventKey.EXT to ext)
        val event = convertToJSONString(eventMap) ?: ""
        val label = convertToJSONString(labelMap) ?: ""
        rtcEngine.sendCustomReportMessage(messageId, category, event, label, 0)
    }

    fun writeLog(content: String, level: Int) {
        rtcEngine.writeLog(level, content)
    }

    fun cleanCache() {
        durationEventStartMap.clear()
    }

    // ---------------------- private ----------------------

    private fun configParameters() {
        //rtcEngine.setParameters("{\"rtc.qos_for_test_purpose\": true}") //测试环境使用
        // 数据上报
        rtcEngine.setParameters("{\"rtc.direct_send_custom_event\": true}")
        // 日志写入
        rtcEngine.setParameters("{\"rtc.log_external_input\": true}")
    }

    private fun getCurrentTs(): Long {
        return System.currentTimeMillis()
    }

    private fun innerReportCostEvent(ts: Long, name: String, cost: Int, ext: Map<String, Any>) {
        Log.d(tag, "reportCostEvent: $name cost: $cost ms ext: $ext")
        writeLog("reportCostEvent: $name cost: $cost ms", Constants.LOG_LEVEL_INFO)
        val eventMap = mapOf(ApiEventKey.TYPE to ApiEventType.COST.value, ApiEventKey.DESC to name)
        val labelMap = mapOf(ApiEventKey.TIMESTAMP to ts, ApiEventKey.EXT to ext)
        val event = convertToJSONString(eventMap) ?: ""
        val label = convertToJSONString(labelMap) ?: ""
        rtcEngine.sendCustomReportMessage(messageId, category, event, label, cost)
    }

    private fun convertToJSONString(dictionary: Map<String, Any>): String? {
        return try {
            JSONObject(dictionary).toString()
        } catch (e: Exception) {
            writeLog("[$tag]convert to json fail: $e dictionary: $dictionary", Constants.LOG_LEVEL_WARNING)
            null
        }
    }
}

File: iOS/APIExample/APIExample/Examples/Advanced/ThirdBeautify/BeautyAPI/Report/APIReporter.m
//
//  APIReporter.m
//  BeautyAPi
//
//  Created by zhaoyongqiang on 2024/4/17.
//

#import "APIReporter.h"

@interface APIReporter ()

@property (nonatomic, strong) AgoraRtcEngineKit *engine;
@property (nonatomic, copy) NSString *category;
@property (nonatomic, strong) NSMutableDictionary<NSString *, NSNumber *> *durationEventStartMap;
@property (nonatomic, copy) NSString *messsageId;

@end

@implementation APIReporter

- (NSString *)messsageId {
    return @"agora:scenarioAPI";
}

- (instancetype)initWithType:(APIType)type version:(NSString *)version engine:(AgoraRtcEngineKit *)engine {
    self = [super init];
    if (self) {
        _category = [NSString stringWithFormat:@"%ld_iOS%@", (long)type, version];
        _engine = engine; 
        [self configParameters];
    }
    return self;
}

- (void)reportFuncEventWithName:(NSString *)name value:(NSDictionary<NSString *, id> *)value ext:(NSDictionary<NSString *, id> *)ext { 
    NSString *content = [NSString stringWithFormat:@"[APIReporter]reportFuncEvent: %@ value: %@ ext: %@", name, value, ext];
    [self debugApiPrint:content];
    
    NSDictionary *eventMap = @{ @"type": @(APIEventTypeAPI), @"desc": name }; 
    NSDictionary *labelMap = @{ @"apiValue": value, @"ts": @([self getCurrentTs]), @"ext": ext};

    NSString *event = [self convertToJSONString:eventMap]; 
    NSString *label = [self convertToJSONString:labelMap];

    [self.engine sendCustomReportMessage:self.messsageId category:self.category event:event label:label value:0];
}

- (void)startDurationEventWithName:(NSString *)name { 
    self.durationEventStartMap[name] = @([self getCurrentTs]);
}

- (void)endDurationEventWithName:(NSString *)name ext:(NSDictionary<NSString *, id> *)ext {
    NSNumber *beginTs = self.durationEventStartMap[name];
    if (!beginTs) {
        return;
    }
    [self.durationEventStartMap removeObjectForKey:name];
    
    NSInteger ts = [self getCurrentTs]; 
    NSInteger cost = ts - beginTs.integerValue;

    [self reportCostEventWithTs:ts name:name cost:cost ext:ext];
}

- (void)reportCostEventWithName:(APICostEvent)name cost:(NSInteger)cost ext:(NSDictionary<NSString *, id> *)ext { [self.durationEventStartMap removeObjectForKey: [self getCostEventWithName:name]];
    
    [self reportCostEventWithTs:[self getCurrentTs] name:[self getCostEventWithName:name] cost:cost ext:ext];
}

- (void)reportCustomEventWithName:(NSString *)name ext:(NSDictionary<NSString *, id> *)ext {
    NSString *content = [NSString stringWithFormat:@"[APIReporter]reportCustomEvent: %@ ext: %@", name, ext];
    [self debugApiPrint:content];
    
    NSDictionary *eventMap = @{ @"type": @(APIEventTypeCustom), @"desc": name };
    NSDictionary *labelMap = @{ @"ts": @([self getCurrentTs]), @"ext": ext };

    NSString *event = [self convertToJSONString:eventMap]; 
    NSString *label = [self convertToJSONString:labelMap];

    [self.engine sendCustomReportMessage:self.messsageId category:self.category event:event label:label value:0];
}

- (NSString *)getCostEventWithName:(APICostEvent)name {
    switch (name) {
        case APICostEventChannelUsage: return @"channelUsage";
        case APICostEventFirstFrameActual: return @"firstFrameActual";
        case APICostEventFirstFramePerceived: return @"firstFramePerceived";
    }
}

- (void)writeLogWithContent:(NSString *)content level:(AgoraLogLevel)level {
    [self.engine writeLog:level content:content];
}

- (void)cleanCache {
    [self.durationEventStartMap removeAllObjects];
}

#pragma mark - Private Methods
- (void)reportCostEventWithTs:(NSInteger)ts name:(NSString *)name cost:(NSInteger)cost ext:(NSDictionary<NSString *, id> *)ext {
    NSString *content = [NSString stringWithFormat:@"[APIReporter]reportCostEvent: %@ cost: %ld ms ext: %@", name, (long)cost, ext];
    [self debugApiPrint:content];
    [self writeLogWithContent:content level:AgoraLogLevelInfo];

    NSDictionary *eventMap = @{ @"type": @(APIEventTypeCost), @"desc": name }; 
    NSDictionary *labelMap = @{ @"ts": @(ts), @"ext": ext };

    NSString *event = [self convertToJSONString:eventMap];
    NSString *label = [self convertToJSONString:labelMap];

    [self.engine sendCustomReportMessage:self.messsageId category:self.category event:event label:label value:cost];
}

- (void)configParameters { 
    [self.engine setParameters:@"{\"rtc.direct_send_custom_event\": true}"];
    [self.engine setParameters:@"{\"rtc.log_external_input\": true}"];
}

- (NSInteger)getCurrentTs { 
    return (NSInteger)([[NSDate date] timeIntervalSince1970] * 1000.0);
}

- (NSString *)convertToJSONString:(NSDictionary *)dictionary {
    NSError *error;
    NSData *jsonData = [NSJSONSerialization dataWithJSONObject:dictionary options:0 error:&error];
    if (!jsonData) {
        [self writeLogWithContent:[NSString stringWithFormat:@"[APIReporter]convert to json fail: %@ dictionary: %@", error, dictionary] level:AgoraLogLevelWarn];
        return nil;
    }
    NSString *jsonString = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
    return jsonString;
}

- (void)debugApiPrint:(NSString *)message {
#if DEBUG
    NSDateFormatter *formatter = [[NSDateFormatter alloc] init];
    formatter.dateFormat = @"yyyy-MM-dd HH:mm:ss.SSS";
    NSString *timeString = [formatter stringFromDate:[NSDate date]];
    NSLog(@"%@ %@", timeString, message);
#endif
}

@end
